Search Block - ability to specify additional search query vars

tomasztunik opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Add ability to control search query vars of the core/search block.

Fixes #40583 

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

This will enable creating core/search block variations that can limit their search results based on query vars provided. Ie. searching only among `post_type=product` for WooCommerce or searching only through a particular category of posts, or a particular author on a WordPress page.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Adds `query` attribute to core/search block and renders hidden inputs in the search block with provided query vars and values. 

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. register new core/search block variation with query attributes

ie. to search only for pages specify `post_type: 'page'` if WooCommerce is available one could use `post_type: 'product'` - it is possible to use all [public query vars](https://codex.wordpress.org/WordPress_Query_Vars#Query_variables) and also custom taxonomies where custom taxonomy name slug is query key and value will be a slug representing custom taxonomy entry.
```
registerBlockVariation( 'core/search', {
   name: 'test/custom-search',
   title: 'Test Search Block',
   category: 'test',
   attributes: {
      query: {
         post_type: 'page'
      }
   }
} );
```

2. Add Test Search Block with inserter
3. Test search and confirm that post_type query parameter is included in browser address bar and that results are limited to given query.

## Screenshots or screencast <!-- if applicable -->

tomasztunik committed:
Add postType attribute to Search Block

tomasztunik committed:
replace core/search postType attr with query object

tomasztunik commented:
Some questions/ideas I have here.

- Should we allow providing raw query param names as query object keys or should we enter them as `camelCase` and convert them to `snake_case` when creating inputs? Thinking if this could possibly be problematic with custom taxonomies or some custom slug names.
- Should we provide any defaults? we could provide `null` values for empty to expose suggested values for query vars and then don't render inputs if there is no value provided. Thinking if this wouldn't pollute the comment strings for block instances on pages/posts/templates.

tomasztunik committed:
update fixtures

tomasztunik committed:
fix php linting errors

tomasztunik committed:
fix bad php autoformat

tomasztunik commented:
Just FYI - opened PR to `@types/wordpress__blocks` to include variations registration API which is required to make the best out of this feature before we introduce the UI (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60059)

tomasztunik committed:
fix conflicts and rabase to current trunk

Mamaduka commented on packages/block-library/src/search/edit.js:
```diff
@@ -216,22 +217,33 @@ export default function SearchEdit( {
+						setAttributes( { placeholder: event.target.value } )
+					}
+				/>
+				{ !! query &&
+					Object.keys( query ).map( ( key ) => (
+						<input
+							key={ key }
+							type="hidden"
+							name={ key }
+							value={ query[ key ] }
+						/>
+					) ) }
```
Hi, @tomasztunik

Sorry if I'm missing something, but what's the reason for outputting `hidden` inputs in the editor?

tomasztunik replied:
Added for consistency in the DOM between the editor/front-end but can be easily removed if we don't see a reason for them to be here.

Mamaduka replied:
Thanks for the explanation, @tomasztunik 🙇 

The hidden inputs shouldn't affect anything visually; they also serve no purpose in the editor besides consistency. So it's probably safe to remove.

tomasztunik replied:
I had something ringing the bell but double checked and they are completely ignored in rendering, for CSS selectors etc. Will remove them 👍

adamziel commented:
In a follow-up PR it could be worth exploring editor controls for this.

tomasztunik commented:
Agreed, in the [issue](https://github.com/WordPress/gutenberg/issues/40583) I've shared my findings that there already is [query-controls](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/query-controls) component but as mentioned there:
> it's not really that generic or doesn't provide freedom as to which options we want to expose nor a simple way to provide autocomplete for custom taxonomies etc.

Having this first would enable us to create variations already and allow more time for the interface to mature at its own pace

Mamaduka commented:
> Having this first would enable us to create variations already and allow more time for the interface to mature at its own pace.

Good point 👍

tomasztunik committed:
Remove rendering hidden query input fields

It didn't add any value to the editor view.
Query attribute is now used only by php
to render the block on the frontend

Mamaduka commented on packages/block-library/src/search/edit.js:
```diff
@@ -216,22 +216,24 @@ export default function SearchEdit( {
-					setAttributes( { placeholder: event.target.value } )
-				}
-			/>
+			<>
```
Small nitpick: We can remove Fragment as well.

tomasztunik replied:
🤦  missed it :)

Mamaduka reviewed (approved):
This looks good to me 👍  Thanks, @tomasztunik!

tomasztunik committed:
Remove rogue <Fragment />

tomasztunik commented:
@Mamaduka and @adamziel - thanks for the quick review 🙇

mburridge commented:
Added Needs Dev Note label in case this needs a dev note for 6.1 release.
